### PR TITLE
Tl detector node

### PR DESCRIPTION
**Pull request by nathansherrit**
Updated the tl_detector node to publish the closest waypoint to the red light
Updated the waypoint_follower node to stop at the red light following the walk through on Udacity

### CODE REVIEW

File: ros/src/tl_detector/tl_detector.py [hunk 1]
```diff
@@ -54,8 +52,21 @@ def __init__(self):
     def pose_cb(self, msg):
         self.pose = msg
 
+	# TODO move to image_cb 
+	# Get the light waypoint and state
+	light_wp, state = self.process_traffic_lights()
+	# If yellow or red publish waypoint
+	# Else publish -1
+	if state == 0 or state == 1:
```
Comment (nyukhalov) [inline ros/src/tl_detector/tl_detector.py]: Let's use `TrafficLight.RED` and `TrafficLight.YELLOW` instead of the magic constants `0` and `1`.

Comment (nyukhalov) [reply ros/src/tl_detector/tl_detector.py]: Also, we could extract the check into a method and reuse it in the `image_cb` method.

File: ros/src/tl_detector/tl_detector.py [hunk 2]
```diff
@@ -131,21 +149,44 @@ def process_traffic_lights(self):
...
+            	if light_stop_wp >= car_position :    #if waypoint is  close to the traffic light and ahead of the car
+            		if first:    #check for the first light 
+				closest_light_stop_wp = light_stop_wp
+                    		closest_light = light
+				first = False
+                	elif light_stop_wp < closest_light_stop_wp:
+                    		closest_light_stop_wp = light_stop_wp  
+                    		closest_light = light
+
+        if ((car_position is not None) and (closest_light_stop_wp is not None)):
```
Comment (nyukhalov) [inline ros/src/tl_detector/tl_detector.py]: `car_position` is always not `None`. We can remove this condition.

File: ros/src/tl_detector/tl_detector.py [hunk 3]
```diff
@@ -131,21 +149,44 @@ def process_traffic_lights(self):
...
+        	dist_to_light = abs(car_position - closest_light_stop_wp)
+
+        if closest_light and dist_to_light < 50: #we check the status of the traffic light if it's within 50 waypoints
```
Comment (nyukhalov) [inline ros/src/tl_detector/tl_detector.py]: if the condition above is true `closest_light` is always not None so we can simplify the code:

```python
if closest_light_stop_wp is not None:
    dist_to_light = abs(car_position - closest_light_stop_wp)
    if dist_to_light < 50: #we check the status of the traffic light if it's within 50 waypoints    
        state = self.get_light_state(closest_light)
        return closest_light_stop_wp, state
return -1, TrafficLight.UNKNOWN
```

File: ros/src/waypoint_updater/waypoint_updater.py [hunk 4]
```diff
@@ -42,24 +44,56 @@ def __init__(self):
...
+	lane = Lane()
+	
+	closest_idx = self.get_nearest_waypoint_idx()
+	farthest_idx = closest_idx + LOOKAHEAD_WPS
+	base_waypoints_1 = self.base_waypoints.waypoints[closest_idx:farthest_idx]
+	
+	# If there is not a red light close to the car
+	# Get the waypoints like normal
+	# Else get the waypoints that stop at the red light
+	if self.stopline_wp_ind == -1 or (self.stopline_wp_ind >= farthest_idx):
```
Comment (nyukhalov) [inline ros/src/waypoint_updater/waypoint_updater.py]: What will happen if `stopline_wp_ind` is bigger than 0 but less than `closest_idx`?
Maybe change the condition to the following to make it more robust?

```python
if self.stopline_wp_ind < closest_idx or self.stopline_wp_ind > farthest_idx:
```

File: ros/src/tl_detector/tl_detector.py (hunk 3 above)
Comment (elias3) [reply ros/src/tl_detector/tl_detector.py]: IMO, the value 50 should be in a variable / const

**Comment by nyukhalov** [comment]
In #11 I added `.editorconfig`.
Could you please ensure this PR includes the change and reformat all the changes files using the settings (most of IDEs support `.editorconfig` out of the box, if no you can install the respective plugin).

**Comment by nyukhalov** [comment]
Updated the code. Please take a look and merge if it looks okay.
The waypoint updater is not perfect yet. But it will be fixed in next PRs.

**Review by elias3 — APPROVED** [review]

File: ros/src/waypoint_updater/waypoint_updater.py (hunk 4 above)
Comment (nathansherrit) [reply ros/src/waypoint_updater/waypoint_updater.py]: the tl_detector should never send a value that is behind the cars current position therefor
stopline_wp_ind is always greater than closest_idx

**Pull request merged**